_posts/2017-11-17-crates-are-not-safe.md

```diff
@@ -0,0 +1,100 @@
+crates.io](https://twitter.com/josh_triplett/status/849089108727222275), we
+survey all of the crates on crates.io and find that 34% (4216/12360) of the
+crates use an `unsafe` block at least once. Of crates that have been downloaded
+10k times or more, 55% (522/944) use `unsafe`. While these ratios are not
+promising, it may be possible to work around the unsafe crates. The problem gets
+worse, however, when we consider dependencies. A crate that does not use
+`unsafe` but relies on a crate that does is effectively unsafe itself. After
+tracking required "normal" dependencies, 77% (9452/12360) of crates are unsafe
+(the number rises to 78% (9624/12360) if optional dependencies are included).
+This means that most of the 12,000-odd crates on crates.io are unsafe!
+
+In Tock we mark the capsules crate with `#![forbid(unsafe_code)]` to ensure that
+no `unsafe` code can creep in. To see if others are doing this as well, we also
+scan crates.io for the use of that flag. We find that only 0.2% (21/12360) of
+crates are explicitly disallowing unsafe code. Because there is no way to use
+cargo or the compiler to enforce that dependencies are safe (although we floated
+the [idea](https://internals.rust-lang.org/t/pre-rfc-cargo-safety-rails/5535)),
+we check to see if any of those explicitly-safe crates have unsafe dependencies.
+We find that all 21 crates have dependencies that use `unsafe`, effectively
+making their use of `#![forbid(unsafe_code)]` moot.
+
+Of course, it is entirely possible that all uses of `unsafe` in crates.io are
+valid, necessary, and actually safe. But validating this, particularly as crates
+are updated, for all dependencies would be a significant challenge. It seems
+that the Rust ecosystem needs a better mechanism for helping developers ensure
+that the code they are using is actually type-safe.
+
+## Usage of the Standard Library is Pervasive
+
+The other challenge for Tock, as an embedded operating system, is the prevalent
+use of the standard library on crates.io. Running on bare-metal, Tock cannot use
+`std::` and only has access to the libraries provided by `core::`. Any
+dependency we would want to use for Tock needs to include `#![no_std]` so that
+the compiler does not try to include the standard library. Again we survey all
+of the crates on crates.io and find that 93% (11488/12360) of crates use the
+standard library (i.e. do not have `#![no_std]`). When considering only crates
+with 10k or more downloads, the ratio does not change (882/944). However, when
+including required dependencies, the number increases to 97% (12023/12360).
+This makes using any crates from crates.io on an embedded platform nearly
+impossible.
+
+Because many of the standard library modules are simply the `core::` crate
+re-exported, not all of these crates actually use functionality not supported on
+embedded platforms. We could (and
+[have](https://github.com/andersk/enum_primitive-rs/pull/14)) issue pull
```

Comment:
I'd remove this sentence and just say that there seems to be a sense that `no_std` is bad. The sentence reads a little bit complain-y and like people expect too much of us (whereas in practice nobody even knows our pain [that we don't even really have yet]).